### ISSUE TITLE
Incorrect warn(unused_assignments) when panicking in an else statement

### ISSUE THREAD

**Issue by Phlosioneer** [rust-lang/rust#57855]
This code:
[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ed20459511c6e8eb1b6ef2371a65e309)
```rust
fn main() {
    let ability: Result<u8, u8> = Ok(8);
    let mut maybe_cost = None;
    if let Ok(cost) = ability {
        maybe_cost = Some(cost);
    } else {
        panic!();
    }
    let cost = maybe_cost.unwrap();
    
    println!("{:?}", cost);
}
```

Produces the incorrect warning:
```
warning: value assigned to `maybe_cost` is never read
 --> src/main.rs:5:13
  |
5 |     let mut maybe_cost = None;
  |             ^^^^^^^^^^
  |
  = note: #[warn(unused_assignments)] on by default
  = help: maybe it is overwritten before being read?
```
It is incorrect because `maybe_cost` is unwrapped, and then printed.

Strangely, this warning goes away if the `panic!();` is commented out. I suspect this has something to do with the control flow analysis? Replacing `panic!();` with `return;` causes the same warning.

Rust version: 1.33.0 nightly

**Comment by jonas-schievink** [comment]
It's not incorrect, the initial None never makes it past the if since it's overwritten in the if, and the function diverges in the else. The unwrap below can never fail.

**Comment by jonas-schievink** [comment]
To clarify what I mean: The compiler is smart enough to figure out that the `None` isn't needed, you can simply remove it:

```rust
fn main() {
    let ability: Result<u8, u8> = Ok(8);
    let mut maybe_cost;
    if let Ok(cost) = ability {
        maybe_cost = Some(cost);
    } else {
        // Comment out this line to fix warning:
        panic!();
    }
    let cost = maybe_cost.unwrap();
    
    println!("{:?}", cost);
}
```

**Comment by Phlosioneer** [comment]
Ooohhhh right, I forgot that rust allows an uninitialized variable if all code paths initialize it.

Sorry!
